### PR TITLE
Feature/generalize to ring

### PR DESCRIPTION
**Pull request by winitzki**
* [x] Use the `spire` library for type classes `Ring` and `Eq`.
* [x] Test on `Expr[Int]` by specializing the type parameter.
* [x] Add a random values test for equality of large polynomials.
* [ ] Update Readme.

### CODE REVIEW

**Comment by winitzki** [comment]
The README should be changed to reflect the latest code.

File: src/main/scala/com/github/symcal/Expr.scala [hunk 1]
```diff
@@ -196,45 +204,46 @@ final case class Var(name: Symbol) extends Expr {
...
-        case (None, _) ⇒ Sum(IntPow(Const(0), d)) // Empty Sum is equivalent to 0.
+        case (None, _) ⇒ Sum(IntPow(Expr.constZero[T], d)) // Empty Sum is equivalent to 0.
         case (Some(head), tail) if tail.isEmpty ⇒ // If x.expand has only one term, we have nothing to expand.
           Sum(IntPow(head, d))
         case _ ⇒ // x.expand has at least 2 terms, need to expand
-          val newMonomials = Expr.getTermCoeffs(xs.length, d.toInt) map {
+          val newMonomials = Expr.getTermCoeffs(xs.length, d.toValue) map {
             case (coeff, powers) ⇒
-              val newTerms = (xs zip powers).map { case (e, i) ⇒ IntPow(e, i) } :+ Const(coeff)
+              val x = 1
```
Comment (hamdi-jenzri) [inline src/main/scala/com/github/symcal/Expr.scala]: not used.

File: src/test/scala/com/github/symcal/ExprIntSpec.scala [hunk 2]
```diff
@@ -156,9 +175,9 @@ class ExprSpec extends FlatSpec with Matchers {
   it should "simplify constants" in {
     (Const(1) - 2 + 3).simplify shouldEqual Const(2)
     (-Const(1) - 2 + 3).simplify shouldEqual Const(0)
-    (-('x - 1)).subs(Var('x), 0).toInt shouldEqual 1
-    (-('x - 1)).diff('x).toInt shouldEqual -1
-    (-(-'x)).simplify shouldEqual ('x: Expr)
+    (-('x - 1)).subs(Var('x), 0).toValue shouldEqual 1
+    (-('x - 1)).diff('x).toValue shouldEqual -1
+    (-(-Var('x))).simplify shouldEqual ('x: Expr[Int])
```
Comment (hamdi-jenzri) [inline src/test/scala/com/github/symcal/ExprIntSpec.scala]: why did you have to add `Var` here? don't we have the implicit `symbol -> Var`?

File: src/main/scala/com/github/symcal/Expr.scala [hunk 3]
```diff
@@ -1,48 +1,50 @@
 package com.github.symcal
 
+import spire.algebra._
+import spire.implicits._
 import scala.language.implicitConversions
 
-sealed trait Expr {
-  def +(x: Expr): Sum = (this, x) match {
+sealed abstract class Expr[T: Ring : Eq] {
```
Comment (hamdi-jenzri) [inline src/main/scala/com/github/symcal/Expr.scala]: what do we need the `Eq` for?

**Pull request merged**